shared/util/codeql/util/Option.qll

```diff
@@ -0,0 +1,41 @@
+/** Provides a module for constructing optional versions of types. */
+
+/** A type with `toString`. */
+signature class TypeWithToString {
+  string toString();
+}
+
+/**
+ * Constructs an `Option` type that is a disjoint union of the given type and an
+ * additional singleton element.
+ */
+module Option<TypeWithToString T> {
+  private newtype TOption =
+    TNone() or
+    TSome(T c)
+
+  /**
+   * An option type. This is either a singleton `None` or a `Some` wrapping the
+   * given type.
+   */
+  class Option extends TOption {
+    /** Gets a textual representation of this element. */
+    string toString() {
+      this = TNone() and result = "(none)"
+      or
+      exists(T c | this = TSome(c) and result = c.toString())
+    }
+
+    /** Gets the wrapped element, if any. */
+    T asSome() { this = TSome(result) }
+  }
+
+  /** The singleton `None` element. */
+  class None extends Option, TNone { }
+
+  /** A wrapper for the given type. */
+  class Some extends Option, TSome { }
+
+  /** Gets the given element wrapped as an `Option`. */
+  Some some(T c) { result = TSome(c) }
```

Comment:
```suggestion
  /** Gets the given element wrapped as an `Option`. */
  Some some(T c) { result = TSome(c) }
  
  /** Gets the singleton `None`. */
  None none() { result = TNone() }
```

Reply:
That doesn't work. The predicate `none()` is a builtin in QL.

Reply:
(And that's why I didn't add it). One can mostly get by with `instanceof MyOption::None`, but I agree it would have been nice to have for symmetry.

Reply:
You can also use `result.isNone()` with the other utility predicate I suggested, that's good enough.